src/lib.rs

```diff
@@ -124,6 +246,317 @@ pub enum DumpError<T> {
 //
 pub type DumpLzss = lzss::Lzss<6, 4, 0x20, { 1 << 6 }, { 2 << 6 }>;
 
+pub fn from_mem(addr: u32, buf: &mut [u8], _meta: bool) -> Result<(), ()> {
+    let src =
+        unsafe { core::slice::from_raw_parts(addr as *const u8, buf.len()) };
+
+    buf.copy_from_slice(src);
+    Ok(())
+}
+
+pub fn to_mem(addr: u32, buf: &[u8]) -> Result<(), ()> {
+    let dest =
+        unsafe { core::slice::from_raw_parts_mut(addr as *mut u8, buf.len()) };
+
+    // dest.copy_from_slice(buf); doesn't work here?!
```

Comment:
That makes me _extremely_ suspicious of UB 😅

Reply:
Thinking a little more, I wonder if `from_raw_parts_mut` is invalid here: it's pointing to a random point in spare SRAM, so it is not necessarily initialized.

I'm not sure how the compiler would prove it, though; I did some experiments with Godbolt and couldn't reproduce the issue.

Reply:
@jgallagher discovered [a thread](https://users.rust-lang.org/t/is-it-ub-to-std-copy-padding-bytes/75993), which suggests this _is_ UB; the fix would be to keep it as a pointer and use `std::ptr::copy_nonoverlapping` instead of casting to a slice.

Reply:
This is really odd in that if I force this function to not be inlined, it can work with `copy_nonoverlapping`.  I pushed an update that has many more comments here in terms of what's going on, but it feels possible if not likely that there is a Rust or LLVM issue here.

Reply:
Cliff actually figured this out in `#oxide-rust` [starting around here](https://matrix.to/#/!uMMSVUNAgMdoTtyJPR:oxide.computer/$_h6yB_H1kTxU0fDJdpmyKKLBN-bygYyoeP3PunydGjU?via=oxide.computer&via=unix.house&via=timecube.club) – it's because the SRAM is mapped as a DEVICE memory, which requires 32-bit alignment (due to its MPU configuration).

Doing this "properly" will require changes to the Hubris build system to properly support `extern` memory regions in tasks.  Luckily, I _also_ need those changes, so I'll be doing that today as part of https://github.com/oxidecomputer/hubris/issues/1223; we can revisit this once that's in place.

Reply:
All resolved -- and done properly thanks to `extern-regions`!